src/coreclr/src/tools/Common/Compiler/CoreRTNameMangler.cs

```diff
@@ -390,11 +391,14 @@ public override Utf8String GetMangledMethodName(MethodDesc method)
 
         private Utf8String GetUnqualifiedMangledMethodName(MethodDesc method)
         {
-            Utf8String mangledName;
-            if (_unqualifiedMangledMethodNames.TryGetValue(method, out mangledName))
-                return mangledName;
+            lock (this)
+            {
+                Utf8String mangledName;
+                if (_unqualifiedMangledMethodNames.TryGetValue(method, out mangledName))
```

Comment:
Same here... if we make _unqualifiedMangledMethodNames a lock-free dictionary, we can avoid locking while looking up. We only need locking while adding new entries